src/calendarworker.cpp

```diff
@@ -553,12 +553,30 @@ void NemoCalendarWorker::loadData(const QList<NemoCalendarData::Range> &ranges,
 
     QMultiHash<QString, NemoCalendarData::Event> events;
     QMultiHash<QString, KDateTime> allDay;
+    bool orphansDeleted = false;
 
     foreach (const KCalCore::Event::Ptr e, list) {
         // The database may have changed after loading the events, make sure that the notebook
         // of the event still exists.
-        if (mStorage->notebook(mCalendar->notebook(e)).isNull())
+        if (mStorage->notebook(mCalendar->notebook(e)).isNull()) {
+            // This may be a symptom of a deeper bug: if a sync adapter (or mkcal)
+            // doesn't delete events which belong to a deleted notebook, then the
+            // events will be "orphan" and need to be deleted.
+            if (mStorage->load(e->uid())) {
+                KCalCore::Incidence::Ptr orphan = mCalendar->incidence(e->uid(), KDateTime());
+                if (!orphan.isNull()) {
+                    if (mCalendar->deleteIncidence(orphan)) {
```

Comment:
Looking at deleteIncidence(), it appears DeleteVisitor calls only deleteEvent(), not deleteEventInstances(), leaving recurrence-id instances behind. Compare to NemoCalendarWorker::deleteAll().